src/intervals/construction.jl

```diff
@@ -0,0 +1,216 @@
+- `m ± r` Midpoint radius representation.
+- `@interval(expr)` Convenience macro replacing all number literal in `expr`
+    by intervals.
+- `I"desc"` Parse the string `"desc"` according to the standard, allowing
+    more flexible syntax and guaranteeing that the typed decimal numbers
+    are included in the interval even if they can not be represented as
+    floating point numbers.
+"""
+struct Interval{T} <: Real
+    lo::T
+    hi::T
+
+    function Interval{T}(a, b) where T
+        a = _normalisezero(a)
+        b = _normalisezero(b)
+        new{T}(T(a, RoundDown), T(b, RoundUp))
+    end
+
+    function Interval{T}(a::T, b::T) where T
+        a = _normalisezero(a)
+        b = _normalisezero(b)
+        new{T}(a, b)
+    end
+end
+
+#= Outer constructors =#
+Interval{T}(a) where T = Interval{T}(a, a)
+Interval(a) = Interval(a, a)
+Interval(a::Tuple) = Interval(a...)
+
+Interval(a::T, b::S) where {T<:AbstractFloat, S} = Interval{promote_type(T, S)}(a, b)
+Interval(a::T, b::S) where {T, S<:AbstractFloat} = Interval{promote_type(T, S)}(a, b)
+Interval(a::T, b::S) where {T<:AbstractFloat, S<:AbstractFloat} = Interval{promote_type(T, S)}(a, b)
+Interval(a::T, b::T) where {T<:AbstractFloat} = Interval{T}(a, b)
+
+Interval(a::T, b::S) where {T, S} = Interval{promote_type(default_bound(), T, S)}(a, b)
+Interval(a::T, b::T) where T = Interval{promote_type(default_bound(), T)}(a, b)
+
+#= Irrational =#
+# Single argument Irrational constructor are in IntervalArithmetic.jl
+# as generated functions need to be define last.
+Interval(a::Irrational, b::T) where {T<:AbstractFloat} = Interval{T}(T(a, RoundDown), b)
+Interval(a::T, b::Irrational) where {T<:AbstractFloat} = Interval{T}(a, T(b, RoundUp))
+
+function Interval(a::Irrational, b::Irrational)
+    T = default_bound()
+    return Interval{T}(a, b)
+end
+
+#= Interval =#
+Interval{T}(x::Interval{T}) where T = x
+Interval{T}(x::Interval) where T = Interval{T}(x.lo, x.hi)
+
+#= Complex =#
+Interval(x::Complex) = Interval(real(x)) + im*Interval(imag(x))
+
+# These definitions has been put there because generated functions must be
+# defined after all methods they use.
+@generated function Interval{T}(x::Irrational) where T
+    res = Interval{T}(x(), x())  # Precompute the interval
+    return :(return $res)  # Set body of the function to return the precomputed result
+end
+
+Interval(x::Irrational) = Interval{default_bound()}(x)
+
+"""
+    interval(a, b)
+
+Create an interval, checking whether [a, b] is a valid `Interval`
+If so, then an `Interval(a, b)` object is returned;
+if not, a warning is printed and the empty interval is returned.
+"""
+function interval(a::T, b::S) where {T<:Real, S<:Real}
+    if !is_valid_interval(a, b)
+        @warn "Invalid input, empty interval is returned"
+        return emptyinterval(promote_type(T, S))
+    end
+
+    return Interval(a, b)
+end
+
+interval(a::Real) = interval(a, a)
+
+const checked_interval = interval
```

Comment:
Shouldn't `checked_interval` be flavor dependent? For instance, if the flavor *allows* to have non-compact intervals (e.g., [1,-2] meaning union([1,Inf], [-Inf,-2]), then `checked_interval` would correspond to `Interval`, right? If `checked_interval` is flavor dependent, how should this be handled?

Reply:
I think if `is_valid_interval` is made flavor-dependent then that would be enough?